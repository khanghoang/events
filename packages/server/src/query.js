/* @flow */
import { createClient, RedisClientType } from './redis-client';

import type { CommitConfig } from '../types/Config.type';
import runLua from './runLua';

const queryLua = `
local from = tonumber(ARGV[1]) + 1;
local to = tonumber(ARGV[2]) or redis.call('HLEN', KEYS[1]);
local newArray = {};

for id=from,to do
  table.insert(newArray, {id, redis.call('hget', KEYS[1], id)});
end

return newArray
`;

export const query = async (
  client: RedisClientType,
  namespc: string,
  ...argv: number[]
) => {
  const array = await runLua(client, queryLua, {
    keys: [`${namespc}::events`],
    argv: argv.map(String),
  });

  return array.map(([id, event]) => ({
    id,
    ...JSON.parse(event),
  }));
};

export default (config: CommitConfig) => {
  const client = createClient(config.redis);

  return async (req: any) => {
    const lastEventId = Number(
      req.headers['Last-Event-ID'] || req.query.lastEventId
    );
    const events = await query(client, config.namespc, lastEventId);

    // we trust the input from commit. otherwise we have to do a try-parse
    return events;
  };
};
