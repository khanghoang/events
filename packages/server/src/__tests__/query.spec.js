/* @flow */
import del from 'redis-functional/del';
import hgetall from 'redis-functional/hgetall';

import query from '../query';
import redisClient from '../redis-client';
import runLua from '../runLua';

describe('query endpoint', () => {
  it('should query', async () => {
    await del(redisClient, 'test-query::events');

    await runLua(
      redisClient,
      `
        for id=1,10 do
          redis.call('HSET', 'test-query::events', id, '{"type": "test", "payload":'..id..'}')
        end

        redis.call('SET', 'test-query::id', 10);
      `,
      { argv: [], keys: [] }
    );

    const req = {
      query: {
        lastEventId: 5,
      },
      headers: {},
    };

    const service = query({
      redis: { url: process.env.REDIS_URL },
      namespc: 'test-query',
    });

    const actual = await service(req);
    expect(actual).toMatchSnapshot();
  });
});
