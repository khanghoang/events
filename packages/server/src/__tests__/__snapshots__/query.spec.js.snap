// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query endpoint should query 1`] = `
Array [
  Object {
    "id": 6,
    "payload": 6,
    "type": "test",
  },
  Object {
    "id": 7,
    "payload": 7,
    "type": "test",
  },
  Object {
    "id": 8,
    "payload": 8,
    "type": "test",
  },
  Object {
    "id": 9,
    "payload": 9,
    "type": "test",
  },
  Object {
    "id": 10,
    "payload": 10,
    "type": "test",
  },
]
`;
